### PR TITLE
API/BUG: only output user-requested columns in votable.parse

### LINKED ISSUE #14943: columns parameter not working as expected in parse_single_table/parse
### Description

When using `parse_single_table` to read a VOTable in an XML file, the parameter `columns` doesn't work as expected. The parameter `columns` is internally passed to the function `parse` which describes `columns` as:

```
columns: sequence of str, optional
    List of field names to include in the output. The default is to include all fields.
```

I'm expecting `parse` to have the same issue. I haven't directly tried it, but I've verified that `columns` is being passed to `parse` appropriately.

### Expected behavior

When passing a list of columns to `columns` only the columns in the list should be present in the output. However, all the columns are present in the output, but the ones that are not in the list are shown as a masked constant (NaN) or as an empty array depending on the type declared in the file.

I'm expecting to get:
```
<VOTable length=2>
  id
int64   
-----
    1
    2
```

But I'm getting:
```
<VOTable length=2>
  id  function_id coefficients
int64    int16       object   
----- ----------- ------------
    1          --           []
    2          --           []
```

### How to Reproduce

```
from astropy.io.votable import parse_single_table
xml_file = 'sample.xml'
_usecols = ['id']
votable = parse_single_table(xml_file, columns=_usecols)
```
The contents of `sample.xml` are:
```
<?xml version="1.0" encoding="UTF-8"?>
<VOTABLE version="1.4" xmlns="http://www.ivoa.net/xml/VOTable/v1.3" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.ivoa.net/xml/VOTable/v1.3 http://www.ivoa.net/xml/VOTable/v1.3" xmlns:spec="http://www.ivoa.net/xml/SpectrumModel/v1.01">
<RESOURCE type="results" utype="u:Type">
<TIMESYS ID="time_frame" refposition="BARYCENTER" timeorigin="24.5" timescale="TCB"/><TABLE>
<FIELD datatype="long" name="id">
<DESCRIPTION>Identifier</DESCRIPTION>
</FIELD>
<FIELD datatype="short" name="function_id">
<DESCRIPTION>Identifier defining the functions</DESCRIPTION>
</FIELD>
<FIELD arraysize="*" datatype="double" name="coefficients">
<DESCRIPTION>Function coefficients</DESCRIPTION>
</FIELD>
<DATA>
<TABLEDATA>
  <TR>
    <TD>1</TD>
    <TD>2</TD>
    <TD>31125.820 3682.205 -1562.585</TD>
  </TR>
  <TR>
    <TD>2</TD>
    <TD>3</TD>
    <TD>3112.820 36822.205 13562.585</TD>
  </TR>
</TABLEDATA>
</DATA>
</TABLE>
</RESOURCE>
</VOTABLE>
```

### Versions

astropy version: '6.0.dev278+g196e3c2d2'

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
Fixes #14943
I took me a couple hours to pass my test *and* all other tests locally, so I'm opening the PR while iron is hot, but I expect I'll probably want to refactor a bit before I open to review.

Since I needed to adjust existing tests which were actively checking for previous behaviour, I'm marking this as an API change, but in my opinion it's borderline a bug fix. In other words, I'm aiming astropy 6.1.0, not 6.0.1.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
I'm not convinced that failures are related. In anywise I don't reproduce them locally. Might have something to do with pytest 8.0.0...

File: astropy/io/votable/tree.py
Comment (eerovaher) [inline astropy/io/votable/tree.py:2741]: Why is the conversion from `range` to `list` necessary?

File: astropy/io/votable/tree.py
Comment (eerovaher) [inline astropy/io/votable/tree.py:3366]: You have defined `self._colnumbers: list[int] | None = None  # uninitialized`. Is there a reason to avoid `self._colnumbers: list[int] = []  # uninitialized`?

File: docs/changes/io.votable/15959.api.rst
Comment (eerovaher) [inline docs/changes/io.votable/15959.api.rst:2]: [The recommended way to read VOTables is through the unified I/O interface](https://docs.astropy.org/en/stable/io/votable/index.html#getting-started), so I'd expect the change log entry to describe how that is affected.

File: astropy/io/votable/tests/test_vo.py
Comment (eerovaher) [inline astropy/io/votable/tests/test_vo.py:313]: Doesn't the following work?
```suggestion
    assert t1.colnames == expected_names
```

**Review by eerovaher — COMMENTED** [review]
I am not doing a thorough review, I'm just commenting on a few things that stood out to me.

**Comment by neutrinoceros** [comment]
@eerovaher This is far from redu for review (I actually left it in the middle of a refactor). I'll answer your questions tomorrow when I'm back at my desk !

File: astropy/io/votable/tree.py
Comment (eerovaher) [reply astropy/io/votable/tree.py:3366]: Never mind, I can see that `[]` would be very different from `None`.

File: astropy/io/votable/tree.py
Comment (neutrinoceros) [reply astropy/io/votable/tree.py:2741]: because `range()` produces a 1-time consumable iterator, but this variable is definitely used more than once.

File: docs/changes/io.votable/15959.api.rst
Comment (neutrinoceros) [reply docs/changes/io.votable/15959.api.rst:2]: rephrased

File: astropy/io/votable/tests/test_vo.py
Comment (neutrinoceros) [reply astropy/io/votable/tests/test_vo.py:313]: it does, thanks ! I forgot about `Table.colnames`

File: astropy/io/votable/tree.py
Comment (neutrinoceros) [reply astropy/io/votable/tree.py]: Oh I didn't know that. Was it always true ?

Comment (eerovaher) [reply astropy/io/votable/tree.py]: As far as I know this was true even for `xrange` in Python 2 (but I can't claim to know Python 2 very well).

I am willing to guess that quite a few people might believe a `range` to be single-use because they know that `range` does lazy evaluation and from that they conclude (mistakenly) that `range` must be a generator.

**Comment by neutrinoceros** [comment]
It took me a while to complete, but I'm almost done with refactoring my initial wonky patch. After I'm sure that the refactor still survives CI, I'll squash my history. Next thing on the list is to make sure that `TableElement._parse_binary` also works correctly when `colnumbers is not None`.

**Comment by neutrinoceros** [comment]
Okay I think this should now be ready for review.

File: astropy/io/votable/tests/test_vo.py
Comment (eerovaher) [inline astropy/io/votable/tests/test_vo.py:292]: The separate list for `ids` could be avoided:
```suggestion
        pytest.param(["string_test"], False, ["string_test"], id="first-col-ids"),
```

File: astropy/io/votable/tree.py
Comment (eerovaher) [inline astropy/io/votable/tree.py:2609]: Why is this brand new code using `list(range(...))`?

Comment (neutrinoceros) [reply astropy/io/votable/tree.py:2609]: type consistency ! if I used a range I'd need to either make the annotation more complicated or change the code again when a type checker is enabled.

Comment (eerovaher) [reply astropy/io/votable/tree.py:2609]: Because the purpose of this pull request is to fix a bug, not to refactor code, then I think type consistency is a good enough argument.

File: astropy/io/votable/tree.py
Comment (neutrinoceros) [reply astropy/io/votable/tree.py]: that would make sense, I probably fell in that trap too.

**Comment by pllim** [comment]
This would affect PyVO world the most, so I am asking a review from Tom D and maybe ping @lmichel too as a stakeholder.

**Comment by lmichel** [comment]
The MIVOT annotation readout in Astropy does not care about the table columns at all. It is a simple extension to the resource parsing that extracts a piece of XML.
The connection between the MIVOT annotations and the data rows (table readout) is done in PyVO (PR [#497](https://github.com/astropy/pyvo/pull/497)). 
Currently our PyVO implementation does not do any column filtering (although it should).
This fix you are proposing shouldn't affect my PyVO code, and if it does, I'll fix it as the PR is not merged.

File: astropy/io/votable/tests/test_vo.py
Comment (eerovaher) [inline astropy/io/votable/tests/test_vo.py:360]: For half the tests `column_ids` only has a single element, so there's little reason to reverse it. It looks like having separate tests for single-element and two-element `column_ids` could simplify parametrization a lot.

Comment (neutrinoceros) [reply astropy/io/votable/tests/test_vo.py:360]: I think there's value in keeping the current parametrisation, which makes it trivial to select test cases related to this (otherwise untested) feature. Honestly I don't really see what problem would be solved by splitting the test in two. Is it okay if I just return early when this part of the test is redundant ?
```patch
diff --git a/astropy/io/votable/tests/test_vo.py b/astropy/io/votable/tests/test_vo.py
index c65189e1f2..712191828d 100644
--- a/astropy/io/votable/tests/test_vo.py
+++ b/astropy/io/votable/tests/test_vo.py
@@ -353,6 +353,9 @@ def test_select_columns_by_name_edge_cases(
     t1 = vot1.to_table(use_names_over_ids=use_names_over_ids)
     assert t1.colnames == expected_names
 
+    if len(column_ids) < 2:
+        return
+
     # columns should be returned in the order they are found, which
     # in the general case isn't the order they are requested
     with np.errstate(over="ignore"):
```

Comment (eerovaher) [reply astropy/io/votable/tests/test_vo.py:360]: If you name the two tests I'm proposing well enough then they can still be easily selected with [`pytest -k ...`](https://docs.pytest.org/en/stable/how-to/usage.html#specifying-which-tests-to-run)

Comment (neutrinoceros) [reply astropy/io/votable/tests/test_vo.py:360]: sure, but then I'll essentially just duplicate the test code entirely. Is it worth it ?

Comment (eerovaher) [reply astropy/io/votable/tests/test_vo.py:360]: For the single-element test the parametrization would be so much simpler that it should be worth it.

Comment (neutrinoceros) [reply astropy/io/votable/tests/test_vo.py:360]: I'm honestly not sure, but I'll push the change you're proposing as a separate commit so we know for sure if we're talking about the same thing.

Comment (eerovaher) [reply astropy/io/votable/tests/test_vo.py:377]: I thought that in the single-element case `column_id` is always the same as `expected_name`, but I just noticed that in one case the `column_id` is `"string_test"`, but the `expected_name` is `"string test"`. So the simplifications I had in mind are not possible.

It is possible to have a single test without reversing `column_ids` in its code by providing the reversed `column_ids` as additional parameters.

Comment (neutrinoceros) [reply astropy/io/votable/tests/test_vo.py]: >  So the simplifications I had in mind are not possible.

I went through the same process while trying to implement your suggestion actually :)

> It is possible to have a single test without reversing column_ids in its code by providing the reversed column_ids as additional parameters.

this seems reasonable to me. Done !

**Comment by neutrinoceros** [comment]
rebased to fix linting and refresh CI

**Review by tomdonaldson — COMMENTED** [review]
I'm satisfied that this should achieve the desired functionality and address #14943.  I appreciate that the test case covered the binary formats.

I only have the one question about the use of `__future__` and will approve once resolved.

File: astropy/io/votable/tree.py
Comment (tomdonaldson) [reply astropy/io/votable/tree.py]: Great, thanks!

**Review by tomdonaldson — APPROVED** [review]
I'm comfortable with this improvement and don't expect any negative surprises for either astropy or pyvo users.  Thanks, @neutrinoceros !

File: astropy/io/votable/tree.py
Comment (pllim) [inline astropy/io/votable/tree.py:2468]: Should the underscore be here?
```suggestion
            del self._all_fields[:]
```

Comment (pllim) [reply astropy/io/votable/tree.py:2468]: I also don't grok why `del` is called here for a bunch of stuff, or how all the attribute access would be affected afterwards.

Comment (neutrinoceros) [reply astropy/io/votable/tree.py:2468]: yup, missing an underscore indeed, thanks !
I spent a couple minutes here trying to understand it myself but frankly I don't think I understood it either when I was working on this. It also seems like the branch is never hit in tests so it might very well be plain broken or out of sync. I would still consider cleaning it out of scope here, as this is meant to be a minimal refactor as far as internals are concerned. What do you think ?

Comment (pllim) [reply astropy/io/votable/tree.py]: Yeah that is fine. Thanks!

File: astropy/io/votable/tree.py
Comment (pllim) [inline astropy/io/votable/tree.py:2539]: Can you please elaborate more here on why this "last minute update" is needed? Wouldn't "all fields" already be defined when a table is read in?

Comment (neutrinoceros) [reply astropy/io/votable/tree.py:2539]: I should have written a better comment; I don't think I understand it better than you do now that 3 months have passed. All I can say for sure now is that removing this makes tests fail, so I know it's still needed. It may have been a band-aid from the start: I don't understand how `self._fields` gets out-of-sync with `self._all_fields`, but it does.
I should also note that right now, I am sick, so maybe I'll get a better shot at explaining it in a couple days, but honestly I'm not sure.

Comment (pllim) [reply astropy/io/votable/tree.py:2539]: @tomdonaldson , do you know why? Are you okay with just merging as-is still? Hope you can advise. Thanks!

Comment (tomdonaldson) [reply astropy/io/votable/tree.py:2539]: Thanks for noticing this.  I should have asked the same question.  

As @neutrinoceros mentioned, this is working around the cases where `self._fields` gets out of sync with `self._all_fields`.  I think the workaround is a little messy, so would be in favor of cleaning it up unless that proves even messier. 

I found 2 places that get these out of sync.  That's probably all of them, but it's worth checking for more if we clean this up:
- `TableElement.add_field()` (new method) only appends to `self.fields` (thus `self._fields`) but doesn't touch `all_fields`.  `add_field()` is only called in `TableElement.from_table()`, but that's definitely called in some tests.
  -  Note that `_register_field()`, which replaces the old `_add_field()`, does update `_all_fields`.
- Some tests call `table.fields.extend()` which would evade any updating of `all_fields` even if `add_fields()` were changed to handle that.

So the workaround may not be necessary if those tests were modified to call `add_field()` and `add_field()` were updated to keep `_all_fields` in sync.  Since I may be missing some case and because we can't prevent people from updating the `fields` list (HomogeneousList) without making it a sequence or other slightly scary change, one could argue that the workaround is actually the safest approach.  I think I would prefer an attempt to remove the workaround and revisit if the details get too messy.

Comment (neutrinoceros) [reply astropy/io/votable/tree.py:2539]: > Some tests call table.fields.extend() which would evade any updating of all_fields even if add_fields() were changed to handle that.

thank you so much for looking into it ! Now that you mention it, I *think* this was my initial motivation for this "last minute update" pattern: if we have tests that do this, my interpretation is that this is something we've supported thusfar and it shouldn't break without a deprecation period. Considering, my approach seems a lot simpler than what it would take to deprecate mutating a list attribute. I would propose to resolve this by extending my comment to include this explanation, what do you think ?

For completeness, I also confirm that the suggested change allows all tests to pass

<details><summary>patch</summary>

```patch
diff --git a/astropy/io/votable/tests/test_converter.py b/astropy/io/votable/tests/test_converter.py
index 35b1cd180c..9cdb7e5651 100644
--- a/astropy/io/votable/tests/test_converter.py
+++ b/astropy/io/votable/tests/test_converter.py
@@ -295,17 +295,15 @@ def test_vararray():
     table = tree.TableElement(votable)
     resource.tables.append(table)
 
-    tabarr = []
     heads = ["headA", "headB", "headC"]
     types = ["char", "double", "int"]
 
     vals = [["A", 1.0, 2], ["B", 2.0, 3], ["C", 3.0, 4]]
     for i in range(len(heads)):
-        tabarr.append(
+        table.add_field(
             tree.Field(votable, name=heads[i], datatype=types[i], arraysize="*")
         )
 
-    table.fields.extend(tabarr)
     table.create_arrays(len(vals))
     for i in range(len(vals)):
         values = tuple(vals[i])
diff --git a/astropy/io/votable/tests/test_vo.py b/astropy/io/votable/tests/test_vo.py
index 3057ede310..37187aa97b 100644
--- a/astropy/io/votable/tests/test_vo.py
+++ b/astropy/io/votable/tests/test_vo.py
@@ -812,12 +812,8 @@ def table_from_scratch():
     resource.tables.append(table)
 
     # Define some fields
-    table.fields.extend(
-        [
-            Field(votable, ID="filename", datatype="char"),
-            Field(votable, ID="matrix", datatype="double", arraysize="2x2"),
-        ]
-    )
+    table.add_field(Field(votable, ID="filename", datatype="char"))
+    table.add_field(Field(votable, ID="matrix", datatype="double", arraysize="2x2"))
 
     # Now, use those field definitions to create the numpy record arrays, with
     # the given number of rows
@@ -861,15 +857,15 @@ def test_build_from_scratch(tmp_path):
     resource.tables.append(table)
 
     # Define some fields
-    table.fields.extend(
-        [
-            tree.Field(
-                votable, ID="filename", name="filename", datatype="char", arraysize="1"
-            ),
-            tree.Field(
-                votable, ID="matrix", name="matrix", datatype="double", arraysize="2x2"
-            ),
-        ]
+    table.add_field(
+        tree.Field(
+            votable, ID="filename", name="filename", datatype="char", arraysize="1"
+        )
+    )
+    table.add_field(
+        tree.Field(
+            votable, ID="matrix", name="matrix", datatype="double", arraysize="2x2"
+        )
     )
 
     # Now, use those field definitions to create the numpy record arrays, with
@@ -983,12 +979,8 @@ def _run_test_from_scratch_example():
     resource.tables.append(table)
 
     # Define some fields
-    table.fields.extend(
-        [
-            Field(votable, name="filename", datatype="char", arraysize="*"),
-            Field(votable, name="matrix", datatype="double", arraysize="2x2"),
-        ]
-    )
+    table.add_field(Field(votable, name="filename", datatype="char", arraysize="*"))
+    table.add_field(Field(votable, name="matrix", datatype="double", arraysize="2x2"))
 
     # Now, use those field definitions to create the numpy record arrays, with
     # the given number of rows
diff --git a/astropy/io/votable/tree.py b/astropy/io/votable/tree.py
index 194ac4d3c3..c31b167f67 100644
--- a/astropy/io/votable/tree.py
+++ b/astropy/io/votable/tree.py
@@ -2536,11 +2536,6 @@ class TableElement(
         of the data columns. Contrary to ``fields``, this property should
         list every field that's available on disk, including deselected columns.
         """
-        # last minute update
-        for field in self._fields:
-            if field not in self._all_fields:
-                self._all_fields.append(field)
-
         return self._all_fields
 
     @property
@@ -2655,6 +2650,7 @@ class TableElement(
 
     def add_field(self, field: Field) -> None:
         self.fields.append(field)
+        self.all_fields.append(field)
 
     def _register_field(self, iterator, tag, data, config, pos) -> None:
         field = Field(self._votable, config=config, pos=pos, **data)
```

</details>

Comment (neutrinoceros) [reply astropy/io/votable/tree.py:2539]: (I also pushed a commit with the one line that is neither a change in tests or a removal of my "last minute update")

File: astropy/io/votable/tree.py
Comment (pllim) [inline astropy/io/votable/tree.py:2468]: I think we have to `del` both to be consistent?
```suggestion
            del self._fields[:]
            del self._all_fields[:]
```

**Review by pllim — COMMENTED** [review]
Thanks. So my understanding is that you will push more commit(s) to undo the "last minute" thingy, correct?

**Comment by neutrinoceros** [comment]
>  So my understanding is that you will push more commit(s) to undo the "last minute" thingy, correct?

As far as I understand, undoing it would constitute an API break, so that's not my preferred route. Depending on @tomdonaldson's reply, I would prefer to just extend my comment to explain why it's needed.

File: astropy/io/votable/tree.py
Comment (neutrinoceros) [reply astropy/io/votable/tree.py:2468]: indeed, thanks !

**Comment by tomdonaldson** [comment]
> > So my understanding is that you will push more commit(s) to undo the "last minute" thingy, correct?
> 
> As far as I understand, undoing it would constitute an API break, so that's not my preferred route. Depending on @tomdonaldson's reply, I would prefer to just extend my comment to explain why it's needed.

I definitely wouldn't advocate (in this PR) a direct API change like making the `fields` property immutable (e.g., making it a sequence).  Since there is no documentation mentioning the impacts of modifying `fields` or providing a recommendation on how modify it, starting to assume the only modifications would be through `add_field()` would be an indirect API change.  More importantly, it would be a little presumptuous and unrealistic.

Since the current workaround ("last minute" thing) protects against any `fields` modifications, I'm OK with keeping it as long as the comment is updated to clearly explain why it's happening.

File: astropy/io/votable/tree.py
Comment (pllim) [inline astropy/io/votable/tree.py:2545]: Sorry, one more thought... Does the order matter, @tomdonaldson ? Just appending might not preserve original order.

File: astropy/io/votable/tree.py
Comment (pllim) [inline astropy/io/votable/tree.py:2596]: Does this count as API change? If so, should mention this new keyword in the change log too?

**Comment by pllim** [comment]
Also need a rebase to pick up RTD fix. Thanks!

File: astropy/io/votable/tree.py
Comment (neutrinoceros) [reply astropy/io/votable/tree.py:2545]: Yes, order matters.
I'll admit that it's not evident that this bit works properly because it relies on *other* parts of the module to be correct too. The intention is that `all_fields` follows the order of on-disk fields and *then* the insertion order of manually added fields. In principle, `self.fields` can only contain a similarly ordered subset of `self.all_fields`, but a bug in how `self.fields` is constructed would cascade into an incorrect ordering in `self.all_fields` too.

Comment (pllim) [reply astropy/io/votable/tree.py:2543]: I am starting to ponder if we should just bite the bullet and fix this properly as Tom pointed out in https://github.com/astropy/astropy/pull/15959#discussion_r1569177870 ... 🤔

Comment (neutrinoceros) [reply astropy/io/votable/tree.py:2543]: Can we agree to explore it in a follow up PR so this patch doesn't get too bloated with deprecation logic ?

File: astropy/io/votable/tree.py
Comment (neutrinoceros) [reply astropy/io/votable/tree.py:2594]: You're right. It also made me realize that the docstring is slightly wrong. Thanks !

File: astropy/io/votable/tree.py
Comment (pllim) [reply astropy/io/votable/tree.py:2543]: I am worried that the non-deterministic order would come back and bite us as a even more subtle bug.

Comment (pllim) [reply astropy/io/votable/tree.py:2543]: But if @tomdonaldson can tell me that my worry is unfounded, then I am okay with merging this and have the other idea as follow-up, though if we do that anyway, then this change would be mostly reverted, so is it worth merging and then reverting? 🤪

Comment (neutrinoceros) [reply astropy/io/votable/tree.py:2543]: But it *is* deterministic ? To be clear, I'm proposing to get the follow up deprecation patch going as soon as this one gets merged, not in an undetermined future, so there won't be much opportunities for this to break or even to stay brittle.

File: astropy/io/votable/tree.py
Comment (neutrinoceros) [reply astropy/io/votable/tree.py:2594]: done !

File: astropy/io/votable/tree.py
Comment (neutrinoceros) [reply astropy/io/votable/tree.py:2543]: Sorry I didn't see your second message

> though if we do that anyway, then this change would be mostly reverted, so is it worth merging and then reverting? 🤪

It can't be reverted until `table.fields.append/extend` is disallowed, which requires a deprecation period

File: astropy/io/votable/tree.py
Comment (tomdonaldson) [inline astropy/io/votable/tree.py:2612]: If this method is called repeatedly with different `colnumbers`, `self.fields` could end up out of order and with fields not in `colnumbers`.  Also, for future protection if we lock down public access to `self.fields`, the append should probably be on `self._fields` (e.g., if later we do something like replace `.append()` with `.add_field()`).

If we do anything to fix this now, it should probably be to totally rebuild `self.fields` from `colnumbers`, but I'm uncomfortable with these side effects.  Ideally creating arrays shouldn't modify any attributes of the table but also ideally the encapsulation (public vs. private) aspects of constructing a table could be cleaned up quite a bit, so I'm hesitant to do anything radical in this PR.

Comment (neutrinoceros) [reply astropy/io/votable/tree.py:2612]: good point. Reconstructing `self._fields` entirely seems safer and isn't really more complicated. How about this ?
```patch
diff --git a/astropy/io/votable/tree.py b/astropy/io/votable/tree.py
index 08dade28db..5df03cdf0e 100644
--- a/astropy/io/votable/tree.py
+++ b/astropy/io/votable/tree.py
@@ -2607,9 +2607,12 @@ class TableElement(
         if colnumbers is None:
             colnumbers = list(range(len(self.all_fields)))
 
-        for i, field in enumerate(self.all_fields):
-            if i in colnumbers and field not in self.fields:
-                self.fields.append(field)
+        new_fields = HomogeneousList(
+            Field,
+            values=[f for i, f in enumerate(self.all_fields) if i in colnumbers],
+        )
+        if new_fields != self.fields:
+            self._fields = new_fields
 
         fields = self.all_fields
```
We could also add a warning if reconstruction is triggered.

Comment (tomdonaldson) [reply astropy/io/votable/tree.py:2612]: That implementation looks reasonable to me.  It's interesting (and fine) that references to `all_fields` ensure that `all_fields` is caught up with anything in `fields` just prior to syncing in the other direction.  (I'm not a huge fan of properties since they make such side-effect non-obvious, but that's my own hangup.)

I can see the motivation for a warning in that it would be an usual and possibly unexpected circumstance, but I also wonder if we could come up with succinct wording that a user could understand and act on.  Unless the reconstruction happens due to a logic error we haven't noticed, the user would be doing some fairly unusual things to trigger this (as we'll see if you add a test case for this), so I think a warning is probably not worth the effort.

Comment (neutrinoceros) [reply astropy/io/votable/tree.py]: I pushed it as is. Thank you so much for your thoughtful review !

Comment (tomdonaldson) [reply astropy/io/votable/tree.py]: Thank you for doing the work and being so responsive to questions and comments!  Could you go ahead and create an issue for the proposed follow-on deprecation patch?

**Review by tomdonaldson — APPROVED** [review]
After the last flurry of activity, I think this is ready again.  Thanks everyone for being so thorough!

**Comment by pllim** [comment]
Thanks,  all!

File: astropy/io/votable/tree.py
Comment (neutrinoceros) [reply astropy/io/votable/tree.py]: It's getting late, but this will be top of my task list for tomorrow !

Comment (pllim) [reply astropy/io/votable/tree.py]: No worries. I opened https://github.com/astropy/astropy/issues/16314 for you. Feel free to add details there tomorrow as you see fit. Thanks!

**Review by msdemlei — COMMENTED** [review]
Disclaimer: I don't know enough about astropy's VOTable implementation to confidently say I could spot subtle issues, and I certainly would not have thought about the problem with the column order in all_fields. Having said that, I agree with this PR's intent, and the implementation is sane and reasonably clear.

I also agree that appending to fields should be deprecated.  For incrementally building table *contents*, there's astropy.Table itself, and I don't see an overriding reason to enable this in VOTable.  There are certainly good reasons to keep VOTable metadata mutable elsewhere (PARAM-s, GROUP-s, INFO-s, etc).  But since mutability is a curse, we should not have it where there are good alternatives.

**Comment by neutrinoceros** [comment]
> I also agree that appending to fields should be deprecated.

@msdemlei maybe you meant to comment directly on #16316 ?

**Comment by msdemlei** [comment]
On Tue, Aug 20, 2024 at 12:10:00AM -0700, Clément Robert wrote:
> > I also agree that appending to fields should be deprecated.
>
> @msdemlei maybe you meant to comment directly on #16316 ?

Not really; I had not bothered to check whether this one was already
merged, and I was asked to review it.  If there's any dissent about
the deprecation, I'll be happy to repeat my endorsement there...

**Comment by pllim** [comment]
Thanks for the comments. Would help for future reference regardless.

**Pull request merged**
